libsolidity/experimental/analysis/TypeInference.cpp

```diff
@@ -628,8 +557,6 @@ bool TypeInference::visit(TypeClassInstantiation const& _typeClassInstantiation)
 		[&](ASTPointer<IdentifierPath> _typeClassName) -> std::optional<TypeClass> {
 			if (auto const* typeClassDefinition = dynamic_cast<TypeClassDefinition const*>(_typeClassName->annotation().referencedDeclaration))
 			{
-				// visiting the type class will re-visit this instantiation
-				typeClassDefinition->accept(*this);
```

Comment:
I'm not 100% I can really remove this.

Initially I assumed that we're doing the separate pass partially to get rid of the necessity to do such a jump. But I see that `visit(TypeClassDefinition)` then recursively goes back to here and I wonder if this has other consequences in the type system. We do not have any tests that would reveal potential problems though so it's hard to tell.

Reply:
Looks like it's still needed. Otherwise type inference breaks if the instantiation is visited before the class:

```solidity
pragma experimental solidity;

type word = __builtin("word");

type T = word;

instantiation T: D {
    function f(self: T) -> word {}
}

class Self: D {
    function f(self: Self) -> word;
}
```
```
TypeError 7428: Type mismatch for function f T -> word != 'x:type -> 'w:type
```

Reply:
We don't need it, if we already construct the full type of the class functions in the previous pass - but yeah, that depends on already constructing types from type expressions, which we still need to disentangle from inference (see also https://github.com/ethereum/solidity/pull/14578/files#r1343054472)